### PR TITLE
Adds `ANY`, `SOME`, and `EVERY` SQL aggregation function tests

### PR DESCRIPTION
**Pull request by alancai98**
Addresses https://github.com/partiql/partiql-tests/issues/79.

Adds the `ANY`, `SOME`, and `EVERY` SQL aggregation function tests from `partiql-lang-kotlin`.

Also does a slight refactor (captured in the first commit https://github.com/partiql/partiql-tests/commit/2b6f95b3f60a55bd9a963aad14d78a75d2415624) of the syntax `COLL_` aggregation tests. The second commit adds the new aggregation function tests.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Review by vgapeyev — COMMENTED** [review]
I highlighted a few places where answers could be different, let's have another look at them.  

I was pretty sure when I was making the comments, but after some thoughts while looking at PR#81, I realize you might have had different guidelines in mind.

File: partiql-tests-data/eval/primitives/coll-aggregate-function.ion [hunk 1]
```diff
@@ -344,6 +344,366 @@ coll_min::[
...
+    statement:"COLL_ANY([ 5, 'hello', 3.14])",
+    assert:[
+      {
+        evalMode: EvalModeError,
+        result:EvaluationFail,
+      },
+      {
+        evalMode:EvalModeCoerce,
+        result:EvaluationSuccess,
+        output:$missing::null
```
[the remark this answers is not in the document]
Comment (alancai98) [reply partiql-tests-data/eval/primitives/coll-aggregate-function.ion:446]: You bring up some good points. I had a few reasons I was outputting `MISSING` rather than `NULL` for these `COLL_` functions.

1. I was defining the `COLL_` functions to follow the general behavior for builtin functions. That is, if there is a mistyped argument, we should return missing for the entire function call in permissive mode (error in strict mode). See [section 7.1](https://partiql.org/assets/PartiQL-Specification.pdf#subsection.7.1) of the spec:

> All functions return `MISSING` when they input data whose types do not conform to the input argument types. 

That's why I believed `MISSING` would be output in this case for the function expression since the input list contained non-boolean, non-null/missing values.

2. Returning `MISSING` in permissive mode can be helpful for the user to know that a data type mismatch had occurred rather than ignoring the mistyped argument.

3. For your point about boolean expressions (e.g. `AND`, `OR`, and also `IN`) following 3-value logic, I defined the `COLL_` following the builtin function semantics rather than the boolean expression semantics. I felt it would be more consistent if all the `COLL_` followed the same data type mismatch behavior.

E.g. `COLL_SUM(['non-integer'] -> MISSING` and `COLL_ANY(['non-boolean', true]) -> MISSING` both returning missing on a mistyped element.

4. Returning `MISSING` for `COLL_ANY`/`COLL_SUM`/`COLL_EVERY` is consistent with the rewrite of SQL aggregates mentioned in [section 11.2.2](https://partiql.org/assets/PartiQL-Specification.pdf#subsubsection.11.2.2).

Consider a PartiQL query as follows:

```
SELECT a, ANY(a) AS agg_result
FROM <<{'a': true}, {'a': false}, {'a': 5}>> AS t
GROUP BY a GROUP AS g
```

The expected result in permissive mode would be:

```
<<
  {
    'a': false,
    'agg_result': true
  },
  {
    'a': true,
    'agg_result': true
  },
  {
    'a': 5,
    -- 'agg_result' omitted since `MISSING` attributes are ignored
  }
>>
```

If we rewrite to use `COLL_ANY` rather than `ANY`, the rewritten query would be something like:
```
SELECT a, COLL_ANY(SELECT VALUE p.t.a FROM g AS p) AS agg_result
FROM <<{'a': true}, {'a': false}, {'a': 5}>> AS t
GROUP BY a GROUP AS g
```

The result of the subquery for the `a = 5` group would be `[5]`. So the result of `COLL_ANY([5])` would have to be `MISSING` in this case.

---

I guess to summarize, I made the following assumptions to determine `MISSING` was returned rather than `NULL`:
- `COLL_` functions all act like other builtin call functions -- return `MISSING` on data type mismatches
- `COLL_ANY`/`COLL_SOME`/`COLL_EVERY` don't follow the boolean expression 3-value logic but are consistent with builtin functions
- Assuming SQL aggregates return `MISSING` on data type mismatches (which is one underspecified aspect part of this spec issue https://github.com/partiql/partiql-spec/issues/50)

File: partiql-tests-data/eval/primitives/coll-aggregate-function.ion [hunk 2]
```diff
@@ -344,6 +344,366 @@ coll_min::[
...
+    statement:"COLL_ANY(<< true, 5, true >>)",
+    assert:[
+      {
+        evalMode: EvalModeError,
+        result:EvaluationFail,
+      },
+      {
+        evalMode:EvalModeCoerce,
+        result:EvaluationSuccess,
+        output:$missing::null
```
[the remark this answers is not in the document]
Comment (alancai98) [reply partiql-tests-data/eval/primitives/coll-aggregate-function.ion:431]: Discussion related to this is captured in the other comment.

**Review by vgapeyev — CHANGES_REQUESTED** [review]
Except for the mis-hap with `AVG([true, false])` and such, this is good to go.  

For clarity, I added comments to retract my previous comments, in accordance with the following. 

To recap our in-person discussions, the rules we settled on for computing COLL_X( c ) are like this: 

- COLL_X(NULL) is NULL, by the general rule of NULL propagating over a function call.
- COLL_X(MISSING) is MISSING, by the general rule of MISSING propagating over a function call.

- COLL_X expects a collection, so if c is not a collection, the result is a type error in strict mode and MISSING in permissive mode. 

- Otherwise, when c is a collection, let c' be its residual with all NULL and MISSING elements removed. 
Then, c' must contain only elements that are of the appropriate type for function X.  
(When X is SUM or AVG the appropriate types are numeric, when X is SOME, ANY, EVERY, the type is boolean.) So, 
   - If c' contains an element of an inappropriate type, the result is a type error in strict mode and MISSING in permissive mode.
   - If c' only contains elements of appropriate types, the result is computed as expected for X.

**Comment by vgapeyev** [comment]
Following up, maybe this is a good opportunity to also include tests for `COLL_X(NULL)`, `COLL_X(MISSING)`, `COLL_X([MISSING])`, and perhaps, `COLL_X(<<MISSING, MISSING>>)` for a good measure?  Not sure whether this needs to be done for all `X`.

File: partiql-tests-data-extended/fail/static-analysis/query/sql-aggregate.ion [hunk 3]
```diff
@@ -5,107 +5,170 @@
...
+        statement:"SUM(ALL [true, true, false, false])",
         assert: {
             result: StaticAnalysisFail
         },
     },
     {
         name:"top level SUM DISTINCT disallowed",
-        statement:"SUM(DISTINCT [1, 2, 3, 3])",
+        statement:"SUM(DISTINCT [true, true, false, false])",
         assert: {
```
[the remark this answers is not in the document]
Comment (alancai98) [reply partiql-tests-data-extended/fail/static-analysis/query/sql-aggregate.ion:107]: Ah good catch. Yeah was a search-and-replace mistake. Will fix

**Review by vgapeyev — APPROVED** [review]
